app/services/oidc_jwt_token_validate.rb

```diff
@@ -0,0 +1,57 @@
+class OidcJwtTokenValidate
+  attr_reader :decoded_token
+  delegate :logger, to: :Rails
+
+  def initialize(jwt_token)
+    @jwt_token = jwt_token
+  end
+
+  def valid_jwt?
+    valid_signature? && type_bearer? && valid_aud?
```

Comment:
Nit: Can we turn this around so `valid_signature?` is last (as it's the most expensive call?)
```suggestion
    type_bearer? && valid_aud? && valid_signature?
```

Reply:
The `valid_signature?` method sets the value of the `@decoded_token` which is then used in the `type_bearer?` method and `valid_aud?` therefore we need to call that first and then the other two.

Reply:
As discussed, you're right and all methods require the decoded token.

I think I'd extract the token to a method like this:

```ruby
  def decoded_token 
    return @decoded_token if defined?(@decoded_token) # we can't use ||= here as the decoded token might be nil as well
    @decoded_token = JWT.decode(@jwt_token, generate_jwks_uri, true, { algorithm: 'RS256' }).first
  end
```

You can then just access `decoded_token` in `valid_signature?`, `type_bearer?` and `valid_aud?`.